File: app/src/main/java/org/andstatus/app/service/QueueData.java
/*
 * Copyright (c) 2016 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.andstatus.app.service;

import android.support.annotation.NonNull;

import org.andstatus.app.context.MyContextHolder;
import org.andstatus.app.timeline.ViewItem;

/**
 * @author yvolk@yurivolkov.com
 */
public class QueueData extends ViewItem<QueueData> implements Comparable<QueueData> {
    public final static QueueData EMPTY = new QueueData(QueueType.UNKNOWN, CommandData.getEmpty());

    @NonNull
    final QueueType queueType;
    @NonNull
    final CommandData commandData;

    static QueueData getNew(@NonNull QueueType queueType, @NonNull CommandData commandData) {
        return new QueueData(queueType, commandData);
    }

    public QueueData(@NonNull QueueType queueType, @NonNull CommandData commandData) {
        this.queueType = queueType;
        this.commandData = commandData;
    }

    @Override
    public long getId() {
        return commandData.hashCode();
    }

    @Override
    public long getDate() {
        return commandData.getCreatedDate();
    }

    public String toSharedSubject() {
        return queueType.getAcronym() + "; "
                + commandData.toCommandSummary(MyContextHolder.get());
    }

    public String toSharedText() {
        return queueType.getAcronym() + "; "
                + commandData.share(MyContextHolder.get());
    }

    @Override
    public int compareTo(@NonNull QueueData another) {
        return -longCompare(getDate(), another.getDate());
    }

    // TODO: Replace with Long.compare for API >= 19
    private static int longCompare(long lhs, long rhs) {
        return lhs < rhs ? -1 : (lhs == rhs ? 0 : 1);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        QueueData queueData = (QueueData) o;
        if (queueType != queueData.queueType) return false;
        return commandData.getCreatedDate() == queueData.commandData.getCreatedDate();
    }

    @Override
    public int hashCode() {
        int result = queueType.hashCode();
        result = 31 * result + (int) (commandData.getCreatedDate() ^ (commandData.getCreatedDate() >>> 32));
        return result;
    }
}

File: app/src/main/java/org/andstatus/app/service/QueueData.kt
/*
 * Copyright (c) 2016 yvolk (Yuri Volkov), http://yurivolkov.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.andstatus.app.service

import org.andstatus.app.context.MyContextHolder
import org.andstatus.app.timeline.ViewItem

/**
 * @author yvolk@yurivolkov.com
 */
class QueueData protected constructor(val queueType: QueueType, val commandData: CommandData) : ViewItem<QueueData>(false, commandData.getCreatedDate()), Comparable<QueueData> {

    override fun getId(): Long {
        return commandData.hashCode().toLong()
    }

    override fun getDate(): Long {
        return commandData.getCreatedDate()
    }

    fun toSharedSubject(): String {
        return (queueType.acronym + "; "
                + commandData.toCommandSummary( MyContextHolder.myContextHolder.getNow()))
    }

    fun toSharedText(): String {
        return (queueType.acronym + "; "
                + commandData.share( MyContextHolder.myContextHolder.getNow()))
    }

    override operator fun compareTo(other: QueueData): Int {
        return -longCompare(getDate(), other.getDate())
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val queueData = other as QueueData
        return if (queueType != queueData.queueType) false
        else commandData.getCreatedDate() == queueData.commandData.getCreatedDate()
    }

    override fun hashCode(): Int {
        var result = queueType.hashCode()
        result = 31 * result + (commandData.getCreatedDate() xor (commandData.getCreatedDate() ushr 32)).toInt()
        return result
    }

    companion object {
        val EMPTY: QueueData = QueueData(QueueType.UNKNOWN, CommandData.EMPTY)
        fun getNew(queueType: QueueType, commandData: CommandData): QueueData {
            return QueueData(queueType, commandData)
        }

        // TODO: Replace with Long.compare for API >= 19
        private fun longCompare(lhs: Long, rhs: Long): Int {
            return if (lhs < rhs) -1 else if (lhs == rhs) 0 else 1
        }
    }
}